### PR TITLE
X.509 Examples

### PR DESCRIPTION
**Pull request by jfhamme-cccs**
Examples for draft-ietf-cose-x509-07.

### CODE REVIEW

**Comment by laurencelundblade** [comment]
I think that x5t and x5chain need to be protected headers. Maybe x5bag.

Ben Kaduk made the argument for x5u on the list, but I believe it extends to x5t and x5chain and maybe some uses of x5bag.

I think the rationale and requirement from [Section 6 from RFC 7515](https://tools.ietf.org/html/rfc7515#page-18) must be applied here.

File: x509-examples/signed-02.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		     "3082019E30820145A003020102021414A4957FD506AA2AAFC669A880032E8C95B87624300A06082A8648CE3D040302302C312A30280603550403132153616D706C6520434F534520436572746966696361746520417574686F726974793020170D3230313230323137323333325A180F32303533313031303137323333325A302C312A30280603550403132153616D706C6520434F534520436572746966696361746520417574686F726974793059301306072A8648CE3D020106082A8648CE3D030107034200047B447C98F731337AFBE3BAC96E793AF12865F3BD56B647A1729764191AE111F3161B4D56FA42F26E1B18DD87F9DB42F4C9168E420E2CE5E2D149648EE0EE5FB4A3433041300F0603551D130101FF040530030101FF300F0603551D0F0101FF04050303070600301D0603551D0E041604141E6FC4D0C0DA004A8427CBBD3FE05A99EA2D2D11300A06082A8648CE3D0403020347003044022006F99B3ACE00007BFB717784DDD230013D8CDCA0BABE20EE00039BEA0898A6D402200FFAF9DE61C1B6BD28BF5DDB1A191E63B22EAD4A69468D5222C487D53C33C204"
+		  ]
```
Comment (laurencelundblade) [inline x509-examples/signed-02.json:24]: My comment here relates to what I think is an ambiguity in X509 COSE.  I don't think it is clear how the verifier is to determine the end-entity certificate to start the verification with.

It should not be by ordering in x5bag. If it is by order in a list, then it should be x5chain.

It should not be by matching kid with the issuer because nothing says you can or should do that.

And per Ben's comments on x5u and JWS design, it should be by a protected header.

**Comment by jfhamme-cccs** [comment]
Thanks for reviewing the PR.  

> I think that x5t and x5chain need to be protected headers. Maybe x5bag.

The examples were generated according to the current v07 of the draft.  In that version (and as of writing, also the editor's copy in GitHub), the descriptions for x5bag, x5t and x5chain contain the following text: "As the contents of this header parameter are untrusted input, the header parameter can be in either the protected or unprotected header bucket."

I believe your comments should be on the I-D, rather than whether the examples in the PR are correct.  But I'll give my opinion on your comments below.

> Ben Kaduk made the argument for x5u on the list, but I believe it extends to x5t and x5chain and maybe some uses of x5bag.

I have read Ben Kaduk's comments on the list regarding x5u.  His argument, I believe, is specific to the lack of "chain of custody" in obtaining the certificate over a URI via an unprotected "coap-not-s".  I do not think this argument is applicable to x5bag or x5chain since these values are assumed to be passed with the COSE message.  I admit that they could actually be inserted/replaced in the COSE message at any time, even after the signature is generated, but this is no different than the CMS SignedData structure (RFC5652) where the included certificates are not covered by the signature.  While SignedData does provide a more explicit reference to the certificate via the SignerIdentifier, the SignerIdentifier is not protected as part of the signed content.  

> I think the rationale and requirement from [Section 6 from RFC 7515](https://tools.ietf.org/html/rfc7515#page-18) must be applied here.

I do not think that Section 6 of RFC7515 requires a certificate to always be a protected header.  The COSE X.509 I-D says that "Certificates obtained from any of these methods MUST still be validated" by PKIX or other trust structure.  Therefore, if the certificate is trusted (by being in a trust store or by verifying the signature and authority via PKIX path validation), these other parameters in the X.509 certificate are indeed integrity protected and meet the requirements of Section 6 of RFC7515.

File: x509-examples/signed-02.json (hunk 1 above)
Comment (jfhamme-cccs) [reply x509-examples/signed-02.json:24]: I completely agree that ordering in x5bag does not determine which certificate is the end-entity certificate.  I believe that this was by design in order to allow a COSE message to transfer additional certificates (e.g. encryption certificates in a sign message) that the recipient may find useful.  The kid is only for information purposes and RFC8152 says that kid is not a security-critical field and is only a hint about which key to use.  Utilizing such a hint may improve processing, but it must not fail if kid is missing or does not correspond to the certificate SubjectName.  Similarly, an application must not fail to process the signature if the end-entity certificate is not the first certificate in the x5bag.  That does not mean that the first certificate must not be the end-entity certificate or that the kid must not match the SubjectName as in the examples I provided. 

Perhaps you are looking for a more extensive test suite?

Comment (laurencelundblade) [reply x509-examples/signed-02.json:24]: I agree with everything you say. :-)

But I don't know how to find the end-entity cert in this example.

I think the problem lies in the specifications, not the example.

**Comment by laurencelundblade** [comment]
Yes, my problem is with the I-D. Sorry for picking on your correct example.

I agree that section 6 of 7515 does not require certificate to always be protected, but it does require them to be protected if any of the fields in the certificate, like key use, are used in the verification.  CMS doesn't have this requirement and thus I think actually has a small security hole by Ben's argument. I believe his argument extends beyond remotely fetched certificates.

Basically COSE x509 should adopt the same rules as JWS does. Right now it does not.
